runtime/browser/ui/desktop/exclusive_access_bubble_views.cc

```diff
@@ -0,0 +1,442 @@
+  if (bubble_view_context_->IsImmersiveModeEnabled()) {
+    // Skip querying the top container height in non-immersive fullscreen
+    // because:
+    // - The top container height is always zero in non-immersive fullscreen.
+    // - Querying the top container height may return the height before entering
+    //   fullscreen because layout is disabled while entering fullscreen.
+    // A visual glitch due to the delayed layout is avoided in immersive
+    // fullscreen because entering fullscreen starts with the top container
+    // revealed. When revealed, the top container has the same height as before
+    // entering fullscreen.
+    top_container_bottom =
+        bubble_view_context_->GetTopContainerBoundsInScreen().bottom();
+  }
+  // |desired_top| is the top of the bubble area including the shadow.
+  int desired_top = kPopupTopPx - view_->border()->GetInsets().top();
+  int y = top_container_bottom + desired_top;
+
+  if (!ignore_animation_state &&
+      animated_attribute_ == ANIMATED_ATTRIBUTE_BOUNDS) {
+    int total_height = size.height() + desired_top;
+    int popup_bottom = animation_->CurrentValueBetween(total_height, 0);
+    int y_offset = std::min(popup_bottom, desired_top);
+    size.set_height(size.height() - popup_bottom + y_offset);
+    y -= y_offset;
+  }
+  return gfx::Rect(gfx::Point(x, y), size);
+}
+
+gfx::Point ExclusiveAccessBubbleViews::GetCursorScreenPoint() {
+  gfx::Point cursor_pos =
+      gfx::Screen::GetScreenFor(
+          bubble_view_context_->GetBubbleAssociatedWidget()->GetNativeView())
+          ->GetCursorScreenPoint();
+  views::View::ConvertPointFromScreen(GetBrowserRootView(), &cursor_pos);
+  return cursor_pos;
+}
+
+bool ExclusiveAccessBubbleViews::WindowContainsPoint(gfx::Point pos) {
+  return GetBrowserRootView()->HitTestPoint(pos);
+}
+
+bool ExclusiveAccessBubbleViews::IsWindowActive() {
+  return bubble_view_context_->GetBubbleAssociatedWidget()->IsActive();
+}
+
+void ExclusiveAccessBubbleViews::Hide() {
+  animation_->SetSlideDuration(kSlideOutDurationMs);
+  animation_->Hide();
+}
+
+void ExclusiveAccessBubbleViews::Show() {
+  animation_->SetSlideDuration(kSlideInDurationMs);
+  animation_->Show();
+}
+
+bool ExclusiveAccessBubbleViews::IsAnimating() {
+  return animation_->is_animating();
+}
+
+bool ExclusiveAccessBubbleViews::CanMouseTriggerSlideIn() const {
+  return !bubble_view_context_->IsImmersiveModeEnabled();
+}
+
+void ExclusiveAccessBubbleViews::Observe(
+    int type,
+    const content::NotificationSource& source,
+    const content::NotificationDetails& details) {
+  UpdateForImmersiveState();
+}
+
+void ExclusiveAccessBubbleViews::OnWidgetVisibilityChanged(
+    views::Widget* widget,
+    bool visible) {
+  UpdateMouseWatcher();
+}
+
+void ExclusiveAccessBubbleViews::ExitExclusiveAccess() {
+  if (bubble_type_ ==
+      EXCLUSIVE_ACCESS_BUBBLE_TYPE_APPLICATION_FULLSCREEN_CLOSE_INSTRUCTION)
+    bubble_view_context_->GetNativeAppViews()->Close();
```

Comment:
if there is a one more pending window (e.g. a  popup) the application will not exit. I think it's better if this user action is propagated to `Runtime` (and then to `Application` so that it is closed) through `NativeAppWindowDelegate`